src/extractor/way_restriction_map.cpp

```diff
@@ -0,0 +1,199 @@
+#include "extractor/way_restriction_map.hpp"
+
+#include <iterator>
+#include <tuple>
+#include <utility>
+
+namespace osrm
+{
+namespace extractor
+{
+
+namespace
+{
+struct FindViaWay
+{
+    bool operator()(const std::tuple<NodeID, NodeID> value,
+                    const TurnRestriction &restriction) const
+    {
+        const auto &way = restriction.AsWayRestriction();
+        return value < std::tie(way.in_restriction.via, way.out_restriction.via);
+    }
+    bool operator()(const TurnRestriction &restriction,
+                    const std::tuple<NodeID, NodeID> value) const
+    {
+        const auto &way = restriction.AsWayRestriction();
+        return std::tie(way.in_restriction.via, way.out_restriction.via) < value;
+    }
+};
+
+} // namespace
+
+WayRestrictionMap::WayRestrictionMap(const std::vector<TurnRestriction> &turn_restrictions)
+{
+    // get all way restrictions
+    const auto extract_restrictions = [this](const auto &turn_restriction) {
+        if (turn_restriction.Type() == RestrictionType::WAY_RESTRICTION)
+        {
+            const auto &way = turn_restriction.AsWayRestriction();
+
+            // so far we can only handle restrictions that are not interrupted
+            if (way.in_restriction.via == way.out_restriction.from &&
+                way.in_restriction.to == way.out_restriction.via)
+                restriction_data.push_back(turn_restriction);
+        }
+    };
+    std::for_each(turn_restrictions.begin(), turn_restrictions.end(), extract_restrictions);
+
+    const auto as_duplicated_node =
```

Comment:
If this function gets inlined into `by_duplicated_node` we could simply use `std::tie(way.in_restriction.via, way.out_restriction.via, way.in_restriction.from)` and avoid the copy over the tuple.